tools/download_utils/download_utils.py

```diff
@@ -0,0 +1,244 @@
+                                 url_filter: Union[Callable[[list], list],
+                                                   None] = None,
+                                 max_itr: int = 3,
+                                 rate_params: dict = {}) -> int:
+    """Attempt to download a list of URLs in multiple iteration.
+        Each iteration attempts to download calls that failed in previous iteration.
+        NOTE: An extra iteration might occour to attempt failed calls using requests library rather than parallel calls.
+
+        Args:
+            url_list: List of URL with metadata dict object.
+            url_api_modifier: Function to attach API key to url.
+            api_key: User's API key provided by US Census.
+            process_and_store: Function to parse, process and store the response to the passed store path.
+            url_filter: Function to filter out URLs that are to be requested.
+            max_itr: Maximum number iterations to be performed.
+                NOTE: An extra iteration might occour to attempt failed calls using requests library rather than parallel calls.
+            rate_params: Dict with parameters to set parallel request rate limits.
+        
+        Returns:
+            Count of url requests that failed.
+    """
+    loop_ctr = 0
+    if not url_filter:
+        url_filter = default_url_filter
+    logging.info('downloading URLs')
+
+    cur_url_list = url_filter(url_list)
+    failed_urls_ctr = len(url_list)
+    prev_failed_ctr = failed_urls_ctr + 1
+    while failed_urls_ctr > 0 and loop_ctr < max_itr and prev_failed_ctr > failed_urls_ctr:
+        prev_failed_ctr = failed_urls_ctr
+        logging.info('downloading URLs iteration:%d', loop_ctr)
+        download_url_list(cur_url_list, url_api_modifier, api_key,
+                          process_and_store, rate_params)
+        cur_url_list = url_filter(url_list)
+        failed_urls_ctr = len(cur_url_list)
+        logging.info('failed request count: %d', failed_urls_ctr)
+        loop_ctr += 1
+    return failed_urls_ctr
+
+
+# req url
+# TODO add back off decorator with aiohttp.ClientConnectionError as the trigger exception,
+#      try except might need to change for decorator to work
+async def fetch(session: Any, cur_url: str, semaphore: Any, limiter: Any,
+                url_api_modifier: Callable[[dict], str], api_key: str,
+                process_and_store: Callable[[Any, str], int]):
+    # async def fetch(session, url, semaphore):
+    """Fetch a single URL in async fashion.
+
+        Args:
+            session: aiohttp ClientSession object.
+            cur_url: URL to be requested.
+            semaphore: asyncio Semaphore object.
+            limiter: AsyncLimiter object.
+            url_api_modifier: Function to attach API key to url.
+            api_key: User's API key provided by US Census.
+            process_and_store: Function to parse, process and store the response to the passed store path.
```

Comment:
Pls add a comment that it catches all exceptions during download and sets status to failed.

Reply:
fixed.